### PR TITLE
[objc][tests] Add objcgen warning and error tests. Fixes issue #104

### PR DESCRIPTION
**Pull request by dalexsoto**
* Adds support for testing error conditions in Xcode projects
  see XcodeBuildErrorTest inside ObjCGenErrWarnTests.cs
* Adds a library to create warning conditions to test objcgen
  and its correponding test (GenWarningTest inside ObjCGenErrWarnTests.cs).

### CODE REVIEW

File: tests/managedwarn/Foo.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿using System;
+namespace managedwarnerr {
+    public class Foo {
+        public Type MyType { get; set; }
+    }
```
Comment (spouliot) [inline tests/managedwarn/Foo.cs]: looks like spaces, not tabs ;-)

Comment (dalexsoto) [inline tests/managedwarn/Foo.cs]: Kill spaces with 🔥 I mean tabs

**Comment by dalexsoto** [comment]
@spouliot My reasoning behind keeping the Xcode project is to delegate the task of figuring out all clang arguments to it, by doing so we can support tvOS, iOS, watchOS, macOS configs, header search paths, etc.  Without worrying about all different configs in our unit tests and also another point is that this is closer to how our customers will integrate our E4K bits. But I am open to suggestions 😄

File: tests/objcgentest/ObjCGenErrWarnTests.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+				using (var msbuildProc = Process.Start (msbuildInfo)) {
+					msbuildProc.WaitForExit ();
+					Assert.That (msbuildProc.ExitCode, Is.Zero, "msbuildProc");
+				}
+
+				var eargs = new List<string> {
+					"-c",
+					$"{Path.Combine (testcaseBaseDir, "DllOutput", csprojName)}.dll",
+					$"-o={Path.Combine (testcaseBaseDir, "E4KOutput")}"
```
Comment (rolfbjarne) [inline tests/objcgentest/ObjCGenErrWarnTests.cs]: If you put the output in a temporary directory, you can skip the cleanup / directory deletion at the end.

You can tell xcodebuild where to put stuff by passing "CONFIGURATION_BUILD_DIR={tmp path}" as a command-line argument (for an example: https://github.com/mono/Embeddinator-4000/blob/75d1de3/tests/objc-cli/Makefile#L73).

**Comment by rolfbjarne** [comment]
We'll need a new Xcode project for every platform (one for iOS, one for tvOS, one for macOS, one for watchOS...), so I'm not sure it's easier than invoking clang directly.

Also personally I find it much harder to figure out how to massage an Xcode project into giving clang the argument I want to give it than constructing the clang command-line manually...

IMHO the only time it makes sense to have an Xcode project is when we want to test the entire integration with Xcode (in this case we're only interested in what clang says about the generated code), or for when we want to be able to debug the generated code with lldb (in this case we won't, because we're not executing the generated code).

**Comment by dalexsoto** [comment]
Since this is two opinions vs one on removing the Xcode project and just use clang 😛

> IMHO the only time it makes sense to have an Xcode project is when we want to test the entire integration with Xcode (in this case we're only interested in what clang says about the generated code), or for when we want to be able to debug the generated code with lldb (in this case we won't, because we're not executing the generated code).

> Question: do we really need the Xcode project ? or could we just call clang and get the warnings out of stdout ?

and I agree on @rolfbjarne's arguments I'll remove the Xcode project and will use Clang instead.

File: tests/objcgentest/ObjCGenErrWarnTests.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+				Console.SetOut (stdout);
+				Console.SetError (stderr);
+			}
+		}
+
+		// We usually are in bin/debug|release
+		static readonly string XcodeFolderPath = $"{AppDomain.CurrentDomain.BaseDirectory}../../xcode";
+		static readonly string MonoMsBuildPath = "/Library/Frameworks/Mono.framework/Versions/Current/Commands/msbuild";
+		static readonly string MonoPath = "/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono";
+		static readonly string ClangPath = "/usr/bin/clang";
```
Comment (chamons) [inline tests/objcgentest/ObjCGenErrWarnTests.cs]: Should this be xcrun clang ?

File: tests/objcgentest/ObjCGenErrWarnTests.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+					$"{Path.Combine (tempWorkingDir, "DllOutput", csprojName)}.dll",
+					$"-o={e4kOutputDir}"
+				};
+			if (config == Configuration.Debug)
+				eargs.Add ("--debug");
+
+			Driver.Main2 (eargs.ToArray ());
+
+			// Sadly no C# 7 yet
+			// (string sdk, string arch, string sdkName, string minVersion) build_info;
```
Comment (chamons) [inline tests/objcgentest/ObjCGenErrWarnTests.cs]: Could just make a struct with a "nice" constructor. I am looking for C#7 though.

File: tests/objcgentest/ObjCGenErrWarnTests.cs (hunk 3 above)
Comment (rolfbjarne) [inline tests/objcgentest/ObjCGenErrWarnTests.cs]: I agree, since we're using `xcrun clang` elsewhere.

**Comment by dalexsoto** [comment]
@rolfbjarne fixed, using xcrun clang now

**Pull request merged**